### PR TITLE
worklist changes for partr (1/3)

### PR DESCRIPTION
**Pull request by vtjnash**
This PR attempts to implement most of the representation changes in kp/partr, while avoiding actually bringing in the multiq code (yet). These changes are actually mostly independent of that change (e.g we could instead adapt that PR to use arrays, then decide whether to make this change later; or we could chose to do this PR, then never merge partr). But these changes that are being proposed over there seem sensible to me as a stepping stone.

TODO:
- [ ] Fix name of `list_deletefirst!`
- [x] Is it worthwhile to try to import more of the sticky-queue logic from partr? We're adding most of the ability to manage it here (and some of the overhead), but likely still don't fully support using Tasks on the worker threads here—can that wait for the partr PR to finish it, or do we want to move the bits from there that would be needed to make it work (prior to merging partr to make it efficient)?

### CODE REVIEW

File: base/linked_list.jl [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            q.head = q.tail = nothing
+        else
+            q.head = val.next::T
+        end
+    else
+        head_next = head.next
+        while head_next !== val
+            head = head_next
+            head_next = head.next
+        end
```
Comment (kpamnany) [inline base/linked_list.jl]: `if head_next === nothing` ... ?

**Comment by kpamnany** [comment]
So this is basically moving task-related code from `event.jl` into `task.jl` and adding `InvasiveLinkedList{T}`? I'm not quite sure of the motivation for this but it looks okay to me barring the single problem I commented on with the `while` loop.

File: base/linked_list.jl (hunk 1 above)
Comment (JeffBezanson) [reply base/linked_list.jl]: The first line of this function should guarantee that `val` is found.

**Comment by JeffBezanson** [comment]
Part 2 merged into this branch.

**Pull request merged**